crates/storage/engines/libmdbx.rs

```diff
@@ -153,37 +175,40 @@ impl StoreEngine for Store {
 
     fn add_receipt(
         &mut self,
-        block_number: BlockNumber,
+        block_hash: BlockHash,
         index: Index,
         receipt: Receipt,
     ) -> Result<(), StoreError> {
-        self.write::<Receipts>((block_number, index), receipt.into())
+        self.write::<Receipts>((block_hash.into(), index), receipt.into())
     }
 
     fn get_receipt(
         &self,
         block_number: BlockNumber,
         index: Index,
     ) -> Result<Option<Receipt>, StoreError> {
-        Ok(self
-            .read::<Receipts>((block_number, index))?
-            .map(|r| r.to()))
+        if let Some(hash) = self.get_block_hash_by_block_number(block_number)? {
+            Ok(self.read::<Receipts>((hash.into(), index))?.map(|b| b.to()))
+        } else {
+            Ok(None)
+        }
     }
 
     fn add_transaction_location(
         &mut self,
         transaction_hash: H256,
-        block_number: BlockNumber,
+        block_hash: BlockHash,
         index: Index,
     ) -> Result<(), StoreError> {
-        self.write::<TransactionLocations>(transaction_hash.into(), (block_number, index))
+        self.write::<TransactionLocations>(transaction_hash.into(), (block_hash.into(), index))
     }
 
     fn get_transaction_location(
         &self,
         transaction_hash: H256,
-    ) -> Result<Option<(BlockNumber, Index)>, StoreError> {
+    ) -> Result<Option<(BlockHash, Index)>, StoreError> {
```

Comment:
I would keep block number here, imagine a transaction is in more than one block. Then it should return the canonical one.

Reply:
I added all locations in the array so that, on request, we just check which one is the canonical and return that one. If no canonical ones are found, None is returned. What do you think? 

If we still want to return the number, we now can, as it's part of the table value. We just delete the final `map` and return both.

Reply:
What is the benefit of returning the hash instead of the number? I think the former is more ambiguous, with the latter it's obvious that we're talking about the canonical one.

Reply:
I updated this to return both now. 

There are some endpoints in `transaction.rs` rpc handler that need both the hash and the number. If you want to fetch the actual transaction it's more efficient to have the hash directly than the number, but you still need the number to build the response.

We would also probably benefit from a `get_transaction_by_location(block_hash, index)`.